## seekfeatures
This notebook demonstrates the use of ``seekfeatures`` using data HCEPD_100K.csv as example.
<br>

``seekfeatures`` is a module that extracts chemical features from input structure in the form of Simplified molecular-input line-entry system (**SMILES_str**). By using open-source package ``mordred``, this module returns more than 1800 chemical 1800 descriptors at high speed. 

<br>``seekfeatures`` contains three functions:<br>
> `seekfeatures.chem_features` extracts chemical information from molecular structures in the form of Smiles_str. <br>
>
> `seekfeatures.missing_value_list` returns a list of missing values with location in the input dataframe.<br>
>
> `seekfeatures.replace_missing` replaces all non-value entries(marked as specific mordred.error.Missing message in the         
                                dataframe with a value 0.

In [1]:
from seekfeatures import *

# other modules needed for this demonstration
import pandas as pd
import numpy as np

<br>

* **`seekfeatures.chem_features`**

In [2]:
# Use 100K data for demonstration
data = pd.read_csv('../Database/HCEPD_100K.csv') 
# Use first 50 data to illustrate the analysis
data50 = data.head(50)
data50.head()

id                                         SMILES_str   stoich_str  \
0   655365          C1C=CC=C1c1cc2[se]c3c4occc4c4nsnc4c3c2cn1  C18H9N3OSSe   
1  1245190  C1C=CC=C1c1cc2[se]c3c(ncc4ccccc34)c2c2=C[SiH2]...  C22H15NSeSi   
2    21847  C1C=c2ccc3c4c[nH]cc4c4c5[SiH2]C(=Cc5oc4c3c2=C1...   C24H17NOSi   
3    65553   [SiH2]1C=CC2=C1C=C([SiH2]2)C1=Cc2[se]ccc2[SiH2]1  C12H12SeSi3   
4   720918       C1C=c2c3ccsc3c3[se]c4cc(oc4c3c2=C1)C1=CC=CC1   C20H12OSSe   

       mass       pce       voc         jsc  e_homo_alpha  e_gap_alpha  \
0  394.3151  5.161953  0.867601   91.567575     -5.467601     2.022944   
1  400.4135  5.261398  0.504824  160.401549     -5.104824     1.630750   
2  363.4903  0.000000  0.000000  197.474780     -4.539526     1.462158   
3  319.4448  6.138294  0.630274  149.887545     -5.230274     1.682250   
4  379.3398  1.991366  0.242119  126.581347     -4.842119     1.809439   

   e_lumo_alpha                                     tmp_smiles_str  
0     -3.444656        C1=CC=C(C1)c1cc2[se]c3c4occc4c4nsnc4c3c2cn1  
1     -3.474074  C1=CC=C(C1)c1cc2[se]c3c(ncc4ccccc34)c2c2=C[SiH...  
2     -3.077368  C1=CC=C(C1)C1=Cc2oc3c(c2[SiH2]1)c1c[nH]cc1c1cc...  
3     -3.548025  C1=CC2=C([SiH2]1)C=C([SiH2]2)C1=Cc2[se]ccc2[Si...  
4     -3.032680       C1=CC=C(C1)c1cc2[se]c3c4sccc4c4=CCC=c4c3c2o1

In [4]:
# input of ChemFeatures should be a list of string
features_df = chem_features(data50['SMILES_str'])

100%|██████████████████████████████████████████████████████████████████████████████████| 50/50 [00:11<00:00,  6.91it/s]
IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



100%|██████████████████████████████████████████████████████████████████████████████████| 50/50 [00:10<00:00,  4.87it/s]


In [5]:
features_df.head()

ABC      ABCGG  nAcid  nBase    SpAbs_A   SpMax_A  SpDiam_A  \
0  20.142136  16.169815      0      0  33.204238  2.616114  5.057172   
1  20.849242  16.133746      0      0  34.755190  2.600656  5.047921   
2  22.889683  17.893320      0      0  37.437727  2.645420  5.160050   
3  13.313708  11.688393      0      0  21.283525  2.481194  4.637583   
4  19.435029  15.989365      0      0  31.670490  2.627835  5.062549   

      SpAD_A   SpMAD_A   LogEE_A  ...      SRW10     TSRW10          MW  \
0  33.204238  1.383510  4.188377  ...  10.460213  77.559765  394.963154   
1  34.755190  1.390208  4.221130  ...  10.493799  77.890806  401.013897   
2  37.437727  1.386582  4.315006  ...  10.639862  82.032579  363.107941   
3  21.283525  1.330220  3.777767  ...   9.782393  66.966647  319.941201   
4  31.670490  1.376978  4.154569  ...  10.429310  77.119188  379.977407   

         AMW  WPath  WPol  Zagreb1  Zagreb2  mZagreb1  mZagreb2  
0  11.968580   1216    42    146.0    185.0  4.611111  5.000000  
1  10.025347   1337    45    150.0    189.0  4.861111  5.250000  
2   8.252453   1635    49    168.0    215.0  5.083333  5.555556  
3  11.426471    430    20     94.0    115.0  3.166667  3.333333  
4  10.856497   1074    39    142.0    181.0  4.361111  4.750000  

[5 rows x 1825 columns]

 <br>

* **`seekfeatures.missing_value_list`**<br>

Not all molecules have the same features as others, so there will be some non-value entries hidden in the dataframe.

In [6]:
missing_list = missing_value_list(features_df)
missing_list[0:6]

* **`seekfeatures.replace_missing`**<br>

For these non-value entries, we assign them to 0, simply indicating that the molecule does not possess that specific feature.

In [9]:
features_df_n = replace_missing(features_df)

In [8]:
features_df_n.head()

ABC      ABCGG  nAcid  nBase    SpAbs_A   SpMax_A  SpDiam_A  \
0  20.142136  16.169815      0      0  33.204238  2.616114  5.057172   
1  20.849242  16.133746      0      0  34.755190  2.600656  5.047921   
2  22.889683  17.893320      0      0  37.437727  2.645420  5.160050   
3  13.313708  11.688393      0      0  21.283525  2.481194  4.637583   
4  19.435029  15.989365      0      0  31.670490  2.627835  5.062549   

      SpAD_A   SpMAD_A   LogEE_A  ...      SRW10     TSRW10          MW  \
0  33.204238  1.383510  4.188377  ...  10.460213  77.559765  394.963154   
1  34.755190  1.390208  4.221130  ...  10.493799  77.890806  401.013897   
2  37.437727  1.386582  4.315006  ...  10.639862  82.032579  363.107941   
3  21.283525  1.330220  3.777767  ...   9.782393  66.966647  319.941201   
4  31.670490  1.376978  4.154569  ...  10.429310  77.119188  379.977407   

         AMW  WPath  WPol  Zagreb1  Zagreb2  mZagreb1  mZagreb2  
0  11.968580   1216    42    146.0    185.0  4.611111  5.000000  
1  10.025347   1337    45    150.0    189.0  4.861111  5.250000  
2   8.252453   1635    49    168.0    215.0  5.083333  5.555556  
3  11.426471    430    20     94.0    115.0  3.166667  3.333333  
4  10.856497   1074    39    142.0    181.0  4.361111  4.750000  

[5 rows x 1825 columns]

In [10]:
#Check if there is missing value after replacement
missing_list = missing_value_list(features_df_n)
missing_list

[]

-------------------------------------------------------------------------------------------------------------------------------
``seekfeatures`` returns a dataframe wit 1825 chemical featrues of all the molecules. User need to select some of the important features from them by using different data cleaning method. 